libs/editor/egui_editor/src/events.rs

```diff
@@ -1,97 +1,306 @@
-use crate::editor::Editor;
-use egui::{Event, Key, PointerButton, Ui, Vec2};
+use crate::debug::DebugInfo;
+use crate::galleys::Galleys;
+use crate::offset_types::DocCharOffset;
+use crate::unicode_segs;
+use crate::unicode_segs::UnicodeSegs;
+use egui::{Event, Key, PointerButton, Vec2};
+use std::iter;
 use std::ops::Range;
+use unicode_segmentation::UnicodeSegmentation;
 
-impl Editor {
-    pub fn key_events(&mut self, ui: &mut Ui) {
-        if !self.galleys.is_empty() {
-            let (galley_idx, _) = self.galley_and_cursor();
-            self.fix_cursor(galley_idx, false);
-        }
+/// represents a modification made as a result of event processing
+#[derive(Debug)]
+enum Modification<'a> {
+    Cursor { cur: Cursor },   // modify the cursor state
+    Insert { text: &'a str }, // insert text at cursor location
+    Delete,                   // delete selection or character before cursor
+    DebugToggle,              // toggle debug overlay
+}
+
+/// processes `events` and returns a boolean pair representing (text_updated, selection_updated)
+pub fn process(
+    events: &[Event], galleys: &Galleys, buffer: &mut Buffer, segs: &mut UnicodeSegs,
+    cursor: &mut Cursor, debug: &mut DebugInfo,
+) -> (bool, bool) {
+    let modifications = calc_modifications(events, segs, galleys, buffer, *cursor);
+    apply_modifications(modifications, buffer, segs, cursor, debug)
+}
+
+fn apply_modifications(
+    mut modifications: Vec<Modification>, buffer: &mut Buffer, segs: &mut UnicodeSegs,
+    cursor: &mut Cursor, debug: &mut DebugInfo,
+) -> (bool, bool) {
+    let mut text_updated = false;
+    let mut selection_updated = false;
 
-        for event in &ui.ctx().input().events {
-            let mut event_matched = true;
-            println!("{:?}", event);
-            match event {
-                Event::Key { key: Key::ArrowRight, pressed: true, modifiers } => {
-                    self.cursor_unprocessed = true;
-                    self.cursor.x_target = None;
-
-                    let (galley_idx, cur_cursor) = self.galley_and_cursor();
-                    if modifiers.shift {
-                        self.cursor.set_selection_origin();
+    let mut cur_cursor = cursor.clone();
+    modifications.reverse();
+    while let Some(modification) = modifications.pop() {
+        match modification {
+            Modification::Cursor { cur } => {
+                cur_cursor = cur;
+                selection_updated = true;
+            }
+            Modification::Insert { text: text_replacement } => {
+                let replaced_text_range = cur_cursor
+                    .selection()
+                    .unwrap_or(Range { start: cur_cursor.pos, end: cur_cursor.pos });
+
+                modify_subsequent_cursors(
+                    replaced_text_range.clone(),
+                    text_replacement,
+                    &mut modifications,
+                    &mut cur_cursor,
+                );
```

Comment:
events are handled multiple at a time without galleys being recomputed in-between; the text you selected and modified may be somewhere else or no longer exist if a prior modification in the same frame moved or deleted it (see unit tests at bottom of file)